hal/common/critical.c

```diff
@@ -304,9 +294,21 @@ uint32_t core_util_atomic_decr_u32(uint32_t *valuePtr, uint32_t delta)
     return newValue;
 }
 
-void *core_util_atomic_decr_ptr(void **valuePtr, ptrdiff_t delta) {
-    return core_util_atomic_decr((uintptr_t)valuePtr, (uintptr_t)delta);
+#endif
+
+
+bool core_util_atomic_cas_ptr(void **ptr, void **expectedCurrentValue, void *desiredValue) {
+    return (void*)core_util_atomic_cas_u32(
+            (uint32_t *)ptr,
+            (uint32_t *)expectedCurrentValue,
+            (uint32_t)desiredValue);
 }
 
-#endif
+void *core_util_atomic_incr_ptr(void **valuePtr, ptrdiff_t delta) {
```

Comment:
stddef.h is included in the related header ([here](https://github.com/mbedmicro/mbed/blob/master/hal/api/critical.h)) since ptrdiff_t is used in the header.

I wasn't sure what type to use for the second argument, so I just went with what the c++ stdlib used for their atomic operators ([here](http://en.cppreference.com/w/cpp/atomic/atomic/operator_arith2)). `ssize_t` is questionably portable, and `ptrdiff_t` seems more accurate than `intptr_t` in this case since you are adding the difference between the pointers.